### PR TITLE
VMware: Add support for multiple datastores (attempt 2)

### LINKED ISSUE #26095: Unable to create a 2-disks vm on different datastores
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
cloud/vmware_guest.py

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/ansible/infra/library', u'/usr/lib/python2.7/site-packages/ara/plugins/modules']
  python version = 2.7.5 (default, Sep 15 2016, 22:37:39) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### CONFIGURATION
N.A.

##### OS / ENVIRONMENT
Red Hat Enterprise Linux 7.3
VMware Vsphere 6.0

##### SUMMARY
When you try to create a 2-disks vm with each disk on a different datastore, both are created on the datastore specified as the first one.

##### STEPS TO REPRODUCE
Create a playbook with the below example and specify 2 different datastores as the _datastore_ and the _disk2_datastore_ variables.

```yaml
- name: Create VM from template
  vmware_guest:
    name: "{{ inventory_hostname }}"

    hostname: "{{ vcenter_hostname }}"
    username: "{{ vcenter_user }}"
    password: "{{ vcenter_pass }}"
    validate_certs: False

    esxi_hostname: "{{ esxhost }}"
    datacenter: "{{ datacenter }}"

    template: '{{ os_family}}{{ os_vers }}{{os_arch}}-template'
    folder: "{{ folder | default('/vm') }}"

    hardware:
      num_cpus: "{{ vm_cpus }}"
      memory_mb: "{{ vm_memory * 1024 | default(1024) }}"

    disk:
      - size_gb: "{{ disk_size | default(10) }}"
        type: thin
        datastore: "{{ datastore }}"
        autoselect_datastore: False
      - size_gb: "{{ disk2_size | default(10) }}"
        type: thin
        datastore: "{{ disk2_datastore }}"
        autoselect_datastore: False

    networks:
      - name: "{{ vlan }}"
        ip: "{{ primary_ip }}"
        netmask: "{{ netmask }}"
        gateway: "{{ gateway }}"
      - name: "{{ nic2_vlan }}"
        ip: "{{ nic2_ip }}"
        netmask: "{{ nic2_netmask }}"

    customization:
      domain: "{{ domain }}"
      dns_servers: "{{ dns_servers }}"

    annotation: "Created with Ansible"
    wait_for_ip_address: true

  register: newvm
  delegate_to: localhost

```

##### EXPECTED RESULTS
The new vm is created with the each disk on the his own datastore

##### ACTUAL RESULTS
The new vm is created with the both disks on the same datastore

```

```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fixes: #26095

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_guest.py
test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml
test/integration/targets/vmware_guest/tasks/main.yml

### CODE REVIEW

**Comment by Akasurde** [comment]
Depends on: https://github.com/ansible/ansible/pull/45953 and https://github.com/ansible/ansible/pull/48618

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @ckotte @jeking3 @Tomrrow9 Could you please review this ?

File: test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    hardware:
+        num_cpus: 1
+        memory_mb: 512
+    disk:
+        - size: 0gb
+          type: thin
+          datastore: "{{ dslist['json'][0] | basename }}"
+        - size: 0gb
+          type: thin
+          datastore: "{{ dslist['json'][1] | basename }}"
```
Comment (jeking3) [inline test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml:89]: It'll be fairly uncommon to split virtual disks to different datastores unless the consumer has very specific requirements for resiliency or performance.

**Comment by jeking3** [comment]
#45953 needs to be merged, then I can finish #48618.  Otherwise...

File: test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml (hunk 1 above)
Comment (Akasurde) [reply test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml:89]: yes, definitely.

**Review by pdellaert — APPROVED** [review]
LGTM

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/102382/87/tests

**Comment by MrBones757** [comment]
Is there any ETA on this change being verified and put onto the devel branch?
I've got a use case where this would be really useful in a production vmware environment.
Thanks for the new feature :)

**Comment by Akasurde** [comment]
@MrBones757 Thanks for the feedback. We are working to get this merged as soon as possible. you can try these changes and let us know if it works for you. This will expedite the merge process. Thanks.

**Comment by Akasurde** [comment]
@MrBones757 Could you please test this in your environment and let me know ? Thanks.

@pdellaert @jeking3 Could you please review this ? Thanks

**Comment by Akasurde** [comment]
@MrBones757 Any news ? Thanks.

**Comment by pukhanov** [comment]
@Akasurde, these changes appear to be working only when creating a new VM. 
However, when creating a VM from a template I get the following error:
```
The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1551890051.38-122224235648190/AnsiballZ_vmware_guest.py", line 113, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1551890051.38-122224235648190/AnsiballZ_vmware_guest.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1551890051.38-122224235648190/AnsiballZ_vmware_guest.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_vmware_guest_payload_zgnY69/__main__.py", line 2685, in <module>
  File "/tmp/ansible_vmware_guest_payload_zgnY69/__main__.py", line 2674, in main
  File "/tmp/ansible_vmware_guest_payload_zgnY69/__main__.py", line 2194, in deploy_vm
  File "/tmp/ansible_vmware_guest_payload_zgnY69/__main__.py", line 1836, in configure_disks
AttributeError: 'NoneType' object has no attribute 'name'
```
It looks like changing line 1826 in `lib/ansible/modules/cloud/vmware/vmware_guest.py` fixes the issue.
```
            if user_specified_datastore:
-                if vm_obj is None:
+                if vm_obj is None or self.params['template']:
```

**Comment by harvey637** [comment]
Hi,
I have also problems with two disks on different datastores, my config is:
```
      disk:
      - autoselect_datastore: true
        datastore: vmx-2_ssd
        size_gb: 40
      - autoselect_datastore: true
        datastore: vmx-2_store
        size_gb: '20'
```

So I expect disk[0] to be 40 GB autoselected on something like "vmx-2_ssd" and disk[1]
with 20 GB autoselected on something like "vmx-2_store". Both datastores are existing, they are
"vmx-2_store4" and "vmx-2_ssd1".

What happens is, the ONLY the autoselected storage of disk[0] is used for both disks.

Inspecting the code in vmware_guest.py I only find construkts like "... self.params['disk'][0] ..."
as in line 1929:

```
 if 'datastore' in self.params['disk'][0] and \
       isinstance(self.params['disk'][0]['datastore'], str) and \
       ds.name.find(self.params['disk'][0]['datastore']) < 0:
       continue
```

which must be wrong as it never asks for the datastore from the second disk, in my
example for "datastore: vmx-2_store".

From my point of view ther must be some loop_counter/index running through 
"self.params('disk'][index][datastore]", not simply alwasy [0].

good luck
cu
Harvey

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 2]
```diff
@@ -1794,6 +1796,64 @@ def configure_disks(self, vm_obj):
             if disk_index == 7:
                 disk_index += 1
```
Comment (goneri) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1798]: This new block (and a bit of the old code after) should be in a new `attach_disk()` function.

File: test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml [hunk 3]
```diff
@@ -0,0 +1,100 @@
+# Test code for the vmware_guest module.
+# Copyright: (c) 2017, Abhijeet Kasurde <akasurde@redhat.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- name: Wait for Flask controller to come up online
```
Comment (goneri) [inline test/integration/targets/vmware_guest/tasks/create_d1_ds2_c1_f0.yml:5]: Could you please convert the test to the new format?

**Review by goneri — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 4]
```diff
@@ -1794,6 +1796,64 @@ def configure_disks(self, vm_obj):
...
+                datastores = self.cache.get_all_objs(self.content, [vim.Datastore])
+                datastores = [x for x in datastores if
+                              self.cache.get_parent_datacenter(x).name == self.params.get('datacenter', None)]
+
+                if datastores is None or len(datastores) == 0:
+                    self.module.fail_json(msg="Unable to find a datastore for disk at "
+                                              "index %s when autoselecting" % disk_index)
+                datastore_freespace = 0
+                for ds in datastores:
+                    if ds.summary.freeSpace > datastore_freespace:
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1824]: Shouldn't ds.summary.freeSpace be compared to disk.size_?b, rather than zero?

**Review by jillr — CHANGES_REQUESTED** [review]

File: test/integration/targets/vmware_guest/tasks/main.yml [hunk 5]
```diff
@@ -7,6 +7,8 @@
 - include: check_mode.yml
 - include: clone_d1_c1_f0.yml
 - include: create_d1_c1_f0.yml
+# VCSIM does not return the path of VMDKs
```
Comment (goneri) [inline test/integration/targets/vmware_guest/tasks/main.yml:10]: Thanks to the last refact, I don't think we should need to comment the test anymore.

**Comment by john-westcott-iv** [comment]
Thanks for this PR, this provides functionality that I need for my provisioning process.

In testing this PR I am finding that the datastore parameter can not be a cluster. When I specify a cluster I get the error message "Unable to find datastore named X specified in disk configuration at index Y". If I specify a datastore and not a cluster it works correctly. The attached [PatchFile.txt](https://github.com/ansible/ansible/files/3149553/PatchFile.txt) allowed me to specify either a datastore or a cluster.

One other thing I noticed is that running a state=present with a changed datastore/cluster does not seem to reconfigure an existing disk. I'm not sure if this is known behavior or a bug with this new code.

**Comment by harvey637** [comment]
Hi,
up to now nothing has improved in the autoselect_datastore thingy with two or more different datastores.
This is part of my configuration:
ansible-playbook 2.9.0.dev0
  config file = /data/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.5 (default, Apr  9 2019, 14:30:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]

...
  invocation:
    module_args:
      annotation: |-
        Testsystem vmware_guest mit 2 Festplatten
        angelegt am 2019-06-03 13:11
      cdrom: {}
      cluster: null
      convert: null
      customization: {}
      customization_spec: null
      customvalues: []
      datacenter: Datacenter2
      datastore: null
      disk:
      - autoselect_datastore: 'true'
        datastore: vmx-2_store
        size_gb: '20'
      - autoselect_datastore: 'true'
        datastore: vmx-2_ssd
        size_gb: '16'
      esxi_hostname: vmx-2
      folder: Datacenter2/vm/Discovered virtual machine
      force: false
      guest_id: centos64Guest
      hardware:
        memory_mb: '8192'
        num_cpus: '2'
        scsi: paravirtual
      hostname: vcenter2
      is_template: false
      linked_clone: false
      name: ansible-test4
      name_match: first
      networks:
      - device_type: vmxnet3
        ip: XXXXXXXXXXX
        mac: XXXXXXXXXXXXXXXXXXX
        name: XX-Server
        netmask: 255.255.255.0
        type: static
      password: VALUE_SPECIFIED_IN_NO_LOG_PARAMETER
      port: 443
      resource_pool: development
      snapshot_src: null
      state: poweredon
      state_change_timeout: 0
      template: null
      use_instance_uuid: false
      username: XXXXXXXXXXXXXXXXXXXXXXXX
      uuid: null
      validate_certs: false
      vapp_properties: []
      wait_for_customization: false
      wait_for_ip_address: true

**Both disks** were created on a **single datastore**: "vmx-2_store4", the existing datastore
for the second disk "vmx-2_ssd2" was completly ignored.

Sorry, I cant see any improvment here. This module is completely useless if different datastores
due to their different costs/performance/size are essential to be used.
cu
Harvey

**Comment by angystardust** [comment]
@harvey637 I'm the original reporter of #26095 and I can say that this issue is still there since ansible 2.3 (I cannot remember if with the old deprecated _vsphere_guest_ module had the same issue).
@Akasurde do you plan to continue working on this patch? I'm available to help testing as usual ;)

**Comment by angystardust** [comment]
@Akasurde any news on this patch?

**Comment by anusha94** [comment]
Hi @Akasurde ,
I tried the below playbook in my local environment.
```
- name: Testing
  hosts: localhost
  gather_facts: no
  tasks:
    - name: create new VMs
      vmware_guest:
        validate_certs: False
        hostname: "{{ server }}"
        username: "{{ username }}"
        password: "{{ password }}"
        name: "New VM3"
        guest_id: centos64Guest
        datacenter: "Sample_DC"
        hardware:
            num_cpus: 1
            memory_mb: 512
        disk:
            - size: 1gb
              type: thin
              datastore: "local-0"
            - size: 2gb
              type: thin
              datastore: "local-2"
        state: poweredoff
        folder: "folder1/"
      register: clone_d1_c1_f0

    - debug: var=clone_d1_c1_f0


    - name: create new VMs again
      vmware_guest:
        validate_certs: False
        hostname: "{{ server }}"
        username: "{{ username }}"
        password: "{{.password }}"
        name: "New VM3"
        guest_id: centos64Guest
        datacenter: "Sample_DC"
        hardware:
            num_cpus: 1
            memory_mb: 512
        disk:
            - size: 1gb
              type: thin
              datastore: "local-0"
            - size: 2gb
              type: thin
              datastore: "local-2"
        state: poweredoff
        folder: "folder1/"
      register: clone_d1_c1_f0_recreate

    - debug: var=clone_d1_c1_f0_recreate


```

**Findings**:

- With your changes, multiple data stores are created

- Without your changes, disks are created on the first datastore only (local-0 in this case).

Let me know if any other scenario needs to be covered.

**Comment by gen2fish** [comment]
Tested with the files, with failure. Add fix by @pukhanov 
```It looks like changing line 1826 in lib/ansible/modules/cloud/vmware/vmware_guest.py fixes the issue.

            if user_specified_datastore:
-                if vm_obj is None:
+                if vm_obj is None or self.params['template']```

And worked as expected;

```
    "invocation": {
        "module_args": {
            "annotation": null,
            "cdrom": {},
            "cluster": "Cluster-Test",
            "convert": null,
            "customization": {},
            "customization_spec": null,
            "customvalues": [],
            "datacenter": "DC-Test,
            "datastore": null,
            "disk": [
                {
                    "datastore": "DATA_01",
                    "size_gb": 80,
                    "type": "eagerzeroedthick"
                },
                {
                    "datastore": "APP_01",
                    "size_gb": 80,
                    "type": "eagerzeroedthick"
                }
```

**Comment by angystardust** [comment]
@goneri @Akasurde is there any hope to see this patch merged anytime soon?

**Comment by MrBones757** [comment]
Hi @Akasurde,

Apologies for never getting back to you re testing of this module.
I have not been working in this space lately...

I have, however, had a chance to test the changes you have made to date as i had to make some changes to my own vmware scripts the other-day and re-discovered this issue again...

When running the latest version, i got the following error:

fatal: [<host> -> localhost]: FAILED! => {"changed": false, "msg": "No vmdk file found for path specified [[<datastore>] <vm-name>/<vm-name>_0.vmdk]"}

The fix noted above by @gen2fish appears to resolves the error for me, though the module does not produce the desired outcome still... one data store is selected instead of multiple as specified in my disk config:

Goal of below: Duplicate a template and create a second template.
vm_disks is generated by a custom filter plugin that follows my company's business rules to autoselect datastores. It correctly conforms to the requirements in the documentation.

    vmware_guest::
      hostname: "hostname"
      username: "username"
      password: "password"
      validate_certs: no
      name: "{{ vm_properties['desired']['name'] }}"
      template: "{{ vm_properties['state']['uuid'] }}"
      datacenter: "{{ vm_properties['desired']['datacenter'] }}"
      cluster: "{{ vm_properties['desired']['cluster'] }}"
      folder: "{{ vm_properties['desired']['folder'] }}"
      disk: "{{ vm_disks }}"
      is_template: yes
      state: present

**Pull request closed without merging**